### PR TITLE
Pruning partitions for predicates based on partitioning columns

### PR DESCRIPTION
**Pull request by soubhik-c**
- Changed ParamLiteral to inherit from Literal instead of containing it.
- Ensured predicate pushdown to datasources happen for pruning to take
effect.
- Using a partitionPruner closure that is used for RowBuffer as well as
ColumnScan

prechecking bug fixes

- handling shuffledependency during reset
- checking CodegenContext references to generate multiple literal term
for the same ParamLiteral
- marking ParamLiteral as foldable=false because ConstantFolding
shouldn't happen during optimization rather everytime new literals are
passed.
- serializing the dataType efficiently and correctly using kryo. (can
    eliminate dataType altogether, todo next)
- handling preparedStatement ParamLiteral constant replace for Smart and
Embedded connector mode due to filter pushdown now happening as
expected.
- removing redundant pruning test

- Putting converted value in SmartConnector mode while executing prepared
statement, filters in smart connector mode now puts in converted to scala
values instead of catalyst values.

- Adding index pruning support and therefore changing it to explicit
ShuffleExchange instead of going through EnsureRequirements so that
numpartitions are aligned with pruned ColScan and base table RowScan.

## Changes proposed in this pull request

(Fill in the changes here)

## Patch testing

- Added new tests specifically verifying buckets used while querying.
- Asserted bucket changes on constant change with tokenization and plan caching.

## ReleaseNotes.txt changes

(Does this change require an entry in ReleaseNotes.txt? If yes, has it been added to it?)

## Other PRs 

(Does this change require changes in other projects- store, spark, spark-jobserver, aqp? Add the links of PR of the other subprojects that are related to this change)

### CODE REVIEW

**Comment by soubhik-c** [comment]
precheckin is clean.

run precheckin

**Comment by soubhik-c** [comment]
build successful.

**Comment by soubhik-c** [comment]
run precheckin

**Comment by soubhik-c** [comment]
run precheckin

**Comment by soubhik-c** [comment]
@kneeraj should we merge ? precheckin is clean other than known issue @rishitesh indicated about UDF ClassNotFound Exception in UserDefinedFunctionsDUnitTest testExecutorHA

```scala
  EXCEPTION: org.apache.spark.SparkException: Job aborted due to stage failure: Task 5 in stage 1655.0 failed 4 times, most recent failure: Lost task 5.3 in stage 1655.0 (TI
D 36428, pnq-soubhikc3): com.esotericsoftware.kryo.KryoException: Unable to find class: IntegerUDF
Serialization trace:
func$1 (org.apache.spark.sql.internal.UDFFunction$$anonfun$makeFunctionBuilder$1$$anonfun$apply$1)
func$2 (org.apache.spark.sql.catalyst.expressions.ScalaUDF$$anonfun$2)
f (org.apache.spark.sql.catalyst.expressions.ScalaUDF)
        at com.esotericsoftware.kryo.util.DefaultClassResolver.readName(DefaultClassResolver.java:160)
        at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:133)
        at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:693)
        at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:118)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:540)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
        at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:540)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
        at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:540)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:813)
        at org.apache.spark.sql.execution.WholeStageCodegenRDD.read(WholeStageCodegenExec.scala:569)
        at com.esotericsoftware.kryo.serializers.DefaultSerializers$KryoSerializableSerializer.read(DefaultSerializers.java:514)
        at com.esotericsoftware.kryo.serializers.DefaultSerializers$KryoSerializableSerializer.read(DefaultSerializers.java:506)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:813)
        at com.twitter.chill.Tuple2Serializer.read(TupleSerializers.scala:41)
        at com.twitter.chill.Tuple2Serializer.read(TupleSerializers.scala:33)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:813)
        at org.apache.spark.serializer.PooledKryoSerializerInstance.deserialize(PooledKryoSerializer.scala:293)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:75)
        at org.apache.spark.scheduler.Task.run(Task.scala:107)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:277)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: IntegerUDF
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at org.apache.spark.executor.SnappyMutableURLClassLoader.loadClassFromJobJar(SnappyExecutor.scala:156)
        at org.apache.spark.executor.SnappyMutableURLClassLoader$$anonfun$loadClass$2$$anonfun$apply$2$$anonfun$apply$3.apply(SnappyExecutor.scala:140)
        at org.apache.spark.executor.SnappyMutableURLClassLoader$$anonfun$loadClass$2$$anonfun$apply$2$$anonfun$apply$3.apply(SnappyExecutor.scala:140)
        at org.apache.spark.executor.SnappyMutableURLClassLoader.loadJar(SnappyExecutor.scala:145)
        at org.apache.spark.executor.SnappyMutableURLClassLoader$$anonfun$loadClass$2$$anonfun$apply$2.apply(SnappyExecutor.scala:140)
        at org.apache.spark.executor.SnappyMutableURLClassLoader$$anonfun$loadClass$2$$anonfun$apply$2.apply(SnappyExecutor.scala:140)
```

**Comment by kneeraj** [comment]
@soubhik-c I think we should. The UDF thing passes if run independently. Please verify that.

**Comment by kneeraj** [comment]
As per latest conversation with you, you will get a short performance gain determination exercise and take QA's help to run some hydra.

**Comment by soubhik-c** [comment]
Regression results are all good and perf benchmark DUnit with 3 servers showed conservatively 50% gain. TPCH queries showed improvements however not sure how its benefiting from partition pruning, I would rather think tokenization is reusing stuffs and showing a false benefit due to multiple consecutive execution of the same query.

**Comment by soubhik-c** [comment]
precheckin is verified to be clean .. merging..

File: core/src/main/scala/org/apache/spark/sql/SnappyParser.scala [hunk 1]
```diff
@@ -642,6 +642,19 @@ class SnappyParser(session: SnappySession)
         ) |
         MATCH ~> UnresolvedAttribute.quoted _
     ) |
+    '{' ~ FN ~ ws ~ identifier ~ '(' ~ (expression * commaSep) ~ ')' ~ ws ~ '}' ~ ws ~> {
+      (f: Any, e: Any) =>
+        f.asInstanceOf[String] match {
+          case f if f.equalsIgnoreCase("TIMESTAMPADD") =>
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/SnappyParser.scala]: @soubhik-c This is not proper. We cannot clutter the parser with such special case handling and set such an example in code (else others will also follow suit). Please add a separate TimestampAdd function instead which will handle SQL_TSI_DAY (and more later if required) and internally call DateAdd as required (or extend latter).

The one and only case of special handling here is for COUNT because that is the only one which can accept COUNT(*) else even that should have been avoided if possible.

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ParamLiteral.scala [hunk 2]
```diff
@@ -146,23 +180,74 @@ case class ParamLiteral(l: Literal, pos: Int) extends LeafExpression {
...
+ * <br><br>
+ *
+ * We try to locate first foldable expression in a query tree such that all its child is foldable
+ * but parent isn't. That way we locate the exact point where an expression is safe to evalute
+ * once instead of evaluating every row.
+ * <br><br>
+ *
+ * Expressions like <code> select c from tab where
+ *  case col2 when 1 then col3 else 'y' end = 22 </code>
+ * like queries doesn't converts literal evaluation into init method.
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ParamLiteral.scala]: "like queries don't convert literal ..."

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ParamLiteral.scala [hunk 3]
```diff
@@ -146,23 +180,74 @@ case class ParamLiteral(l: Literal, pos: Int) extends LeafExpression {
...
+  override def eval(input: InternalRow): Any = expr.eval(input)
+
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val eval = expr.genCode(ctx)
+    val newVar = ctx.freshName("paramLiteralExpr")
+    val newVarIsNull = ctx.freshName("paramLiteralExprIsNull")
+    val comment = ctx.registerComment(expr.toString)
+    ctx.addMutableState(ctx.javaType(expr.dataType), newVar,
+      s"$comment\n${eval.code}\n$newVar = ${eval.value};")
+    ctx.addMutableState("boolean", newVarIsNull, s"$newVarIsNull = ${eval.isNull};")
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ParamLiteral.scala]: I think this code for both newVar and newVarIsNull should be put inside a separate function instead of cluttering the init method. We also have a problem of 64K method limit so that will be better since "eval.code" can potentially be many and long.

File: core/src/main/scala/org/apache/spark/sql/execution/ExistingPlans.scala [hunk 4]
```diff
@@ -233,13 +242,17 @@ private[sql] final case class ZipPartitionScan(basePlan: CodegenSupport,
   private val consumedVars: ArrayBuffer[ExprCode] = ArrayBuffer.empty
   private val inputCode = basePlan.asInstanceOf[CodegenSupport]
 
-  override def children: Seq[SparkPlan] = basePlan :: otherPlan :: Nil
+  private val withShuffle = ShuffleExchange(HashPartitioning(
+    ClusteredDistribution(otherPartKeys)
+        .clustering, inputCode.inputRDDs().head.getNumPartitions), otherPlan)
+
+  override def children: Seq[SparkPlan] = basePlan :: withShuffle :: Nil
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/ExistingPlans.scala]: Why is this always a shuffle?? What if the partitioning of the two match (which is the more common case)? Either way we discussed that this will be handled automatically by EnsureRequirements and this should not be done explicitly. I guess you will need this if the plan is being used after EnsureRequirements but then it should use a similar way of determining whether to do shuffle, broadcast or collocate join like done by EnsureRequirements.

File: core/src/main/scala/org/apache/spark/sql/execution/ExistingPlans.scala [hunk 5]
```diff
@@ -233,13 +242,17 @@ private[sql] final case class ZipPartitionScan(basePlan: CodegenSupport,
...
 
   override def requiredChildDistribution: Seq[Distribution] =
     ClusteredDistribution(basePartKeys) :: ClusteredDistribution(otherPartKeys) :: Nil
 
   override def inputRDDs(): Seq[RDD[InternalRow]] =
-    inputCode.inputRDDs ++ Some(otherPlan.execute())
+    inputCode.inputRDDs ++ Some(withShuffle.execute())
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/ExistingPlans.scala]: Why always shuffle? This is a big performance bug.

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala [hunk 6]
```diff
@@ -548,7 +592,11 @@ class ColumnFormatRelation(
       df.write.insertInto(snappySession.getIndexTable(indexIdent).toString())
     } catch {
       case e: Throwable =>
-        snappySession.dropTable(indexIdent, ifExists = false)
+        try {
+          snappySession.dropTable(indexIdent, ifExists = false)
+        } catch {
+          case _: Throwable =>
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala]: What is this?? Please never ignore Throwable like this. It hides completely legitimate bugs and also fatal system errors. At most match NonFatal(e).

But why is that required in the first place? If an unexpected exception is coming here then it is a bug in dropTable which should be fixed instead.

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/JDBCSourceAsColumnarStore.scala [hunk 7]
```diff
@@ -320,9 +327,44 @@ class JDBCSourceAsColumnarStore(override val connProperties: ConnectionPropertie
...
+      val region = Misc.getRegionForTable(resolvedName, true)
+      partitionPruner match {
+        case -1 if allPartitions != null =>
+          allPartitions
+        case -1 =>
+          allPartitions = session.sessionState.getTablePartitions(
+            region.asInstanceOf[PartitionedRegion])
+          allPartitions
+        case bucketId: Int =>
+          if (java.lang.Boolean.getBoolean("DISABLE_PARTITION_PRUNING")) {
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/JDBCSourceAsColumnarStore.scala]: Change this system property to a proper conf variable as mentioned in a comment above.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/StoreDataSourceStrategy.scala [hunk 8]
```diff
@@ -202,4 +205,137 @@ private[sql] object StoreDataSourceStrategy extends Strategy {
         filterCondition.map(execution.FilterExec(_, scan)).getOrElse(scan))
     }
   }
+
+  /**
+   * Tries to translate a Catalyst [[Expression]] into data source [[Filter]].
+   *
+   * @return a `Some[Filter]` if the input [[Expression]] is convertible, otherwise a `None`.
+   */
+  protected[sql] def translateFilter(predicate: Expression): Option[Filter] = {
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/StoreDataSourceStrategy.scala]: I suppose this is a copy of some Spark code? Mention the same in the comment of this method "adapted from Spark's ..." and also add the Spark copyright header to this file.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/StoreDataSourceStrategy.scala [hunk 9]
```diff
@@ -202,4 +205,137 @@ private[sql] object StoreDataSourceStrategy extends Strategy {
...
+   * and can be handled by `relation`.
+   *
+   * @return A pair of `Seq[Expression]` and `Seq[Filter]`. The first element contains all Catalyst
+   *         predicate [[Expression]]s that are either not convertible or cannot be handled by
+   *         `relation`. The second element contains all converted data source [[Filter]]s that
+   *         will be pushed down to the data source.
+   */
+  protected[sql] def selectFilters(
+      relation: BaseRelation,
+      predicates: Seq[Expression]): (Seq[Expression], Seq[Filter]) = {
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/StoreDataSourceStrategy.scala]: same comment as above

File: core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala [hunk 10]
```diff
@@ -102,7 +103,32 @@ class SnappySessionState(snappySession: SnappySession)
...
+        p
+    } transform {
+      case q: LogicalPlan => q transformExpressionsDown {
+        // ignore leaf ParamLiteral & Literal
+        case p: ParamLiteral => p
+        case l: Literal => l
+        // Wrap expressions that are foldable.
+        case e if e.foldable =>
+          // lets mark child params foldable false so that nested expression doesn't
+          // attempt to wrap.
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala]: Can you explain this with a short example?
